File: ios/PulseWatchKit/Sources/PulseWatchVault/EnvelopeCipher.swift
import Foundation
import CryptoKit

public enum EnvelopeCipherError: Error {
    case invalidPublicKey
    case sealFailed(Error)
}

public enum EnvelopeCipher {
    static let info = Data("pulsewatch v1".utf8)

    /// Wraps `plaintext` for delivery to `recipientRawPublicKey` (32-byte X25519 raw).
    /// Wire format matches backend `decapsulate`:
    ///   `eph_pub(32) || iv(12) || tag(16) || ciphertext`
    public static func seal(plaintext: Data, recipientRawPublicKey: Data) throws -> Data {
        guard recipientRawPublicKey.count == 32,
              let recipient = try? Curve25519.KeyAgreement.PublicKey(rawRepresentation: recipientRawPublicKey)
        else { throw EnvelopeCipherError.invalidPublicKey }

        let ephemeral = Curve25519.KeyAgreement.PrivateKey()
        let shared: SharedSecret
        do {
            shared = try ephemeral.sharedSecretFromKeyAgreement(with: recipient)
        } catch {
            throw EnvelopeCipherError.sealFailed(error)
        }

        let key = shared.hkdfDerivedSymmetricKey(
            using: SHA256.self,
            salt: recipientRawPublicKey,
            sharedInfo: info,
            outputByteCount: 32
        )

        let nonce = AES.GCM.Nonce()
        let sealed: AES.GCM.SealedBox
        do {
            sealed = try AES.GCM.seal(plaintext, using: key, nonce: nonce)
        } catch {
            throw EnvelopeCipherError.sealFailed(error)
        }

        var out = Data()
        out.reserveCapacity(32 + 12 + 16 + plaintext.count)
        out.append(ephemeral.publicKey.rawRepresentation)
        out.append(Data(nonce))
        out.append(sealed.tag)
        out.append(sealed.ciphertext)
        return out
    }

    /// Test helper — same logic as the backend, useful for unit tests.
    public static func open(wrapped: Data, recipientPrivateKey: Curve25519.KeyAgreement.PrivateKey) throws -> Data {
        guard wrapped.count >= 32 + 12 + 16 else { throw EnvelopeCipherError.invalidPublicKey }
        let ephRaw = wrapped.prefix(32)
        let nonceData = wrapped.dropFirst(32).prefix(12)
        let tag = wrapped.dropFirst(32 + 12).prefix(16)
        let ct = wrapped.dropFirst(32 + 12 + 16)

        let eph = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: ephRaw)
        let shared = try recipientPrivateKey.sharedSecretFromKeyAgreement(with: eph)
        let key = shared.hkdfDerivedSymmetricKey(
            using: SHA256.self,
            salt: recipientPrivateKey.publicKey.rawRepresentation,
            sharedInfo: info,
            outputByteCount: 32
        )
        let nonce = try AES.GCM.Nonce(data: nonceData)
        let box = try AES.GCM.SealedBox(nonce: nonce, ciphertext: ct, tag: tag)
        return try AES.GCM.open(box, using: key)
    }
}

File: android/core/src/main/java/app/pulsewatch/core/vault/EnvelopeCipher.kt
package app.pulsewatch.core.vault

import org.bouncycastle.crypto.agreement.X25519Agreement
import org.bouncycastle.crypto.digests.SHA256Digest
import org.bouncycastle.crypto.generators.HKDFBytesGenerator
import org.bouncycastle.crypto.params.HKDFParameters
import org.bouncycastle.crypto.params.X25519PrivateKeyParameters
import org.bouncycastle.crypto.params.X25519PublicKeyParameters
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * Wire format (matches backend `decapsulate` and iOS `EnvelopeCipher`):
 *   `eph_pub(32) || iv(12) || tag(16) || ciphertext`
 *
 * HKDF salt is the recipient public key, info is the constant `pulsewatch v1`.
 */
object EnvelopeCipher {
    private const val PUB_LEN = 32
    private const val IV_LEN = 12
    private const val TAG_LEN = 16
    private val INFO = "pulsewatch v1".toByteArray(Charsets.UTF_8)
    private val secureRandom = SecureRandom()

    /** Encrypts `plaintext` for [recipientRawPublicKey] (32-byte X25519 raw). */
    fun seal(plaintext: ByteArray, recipientRawPublicKey: ByteArray): ByteArray {
        require(recipientRawPublicKey.size == PUB_LEN) { "recipient pubkey must be 32 bytes" }

        val ephPriv = X25519PrivateKeyParameters(secureRandom)
        val ephPubBytes = ByteArray(PUB_LEN).also { ephPriv.generatePublicKey().encode(it, 0) }

        val shared = ByteArray(X25519Agreement.SECRET_SIZE)
        X25519Agreement().apply { init(ephPriv) }
            .calculateAgreement(X25519PublicKeyParameters(recipientRawPublicKey, 0), shared, 0)

        val key = hkdfSha256(shared, salt = recipientRawPublicKey, info = INFO, length = 32)

        val iv = ByteArray(IV_LEN).also { secureRandom.nextBytes(it) }
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(TAG_LEN * 8, iv))
        val ctAndTag = cipher.doFinal(plaintext)

        // JCA returns ciphertext || tag; we want tag before ciphertext on the wire.
        val ctLen = ctAndTag.size - TAG_LEN
        val ct = ctAndTag.copyOfRange(0, ctLen)
        val tag = ctAndTag.copyOfRange(ctLen, ctAndTag.size)

        // zero the derived key — best-effort on the JVM
        java.util.Arrays.fill(key, 0)
        java.util.Arrays.fill(shared, 0)

        return ephPubBytes + iv + tag + ct
    }

    /** Test helper: opens a payload with the same recipient private key. */
    internal fun open(wrapped: ByteArray, recipientPrivateKey: ByteArray): ByteArray {
        require(wrapped.size >= PUB_LEN + IV_LEN + TAG_LEN) { "wrapped too short" }
        val ephPub = wrapped.copyOfRange(0, PUB_LEN)
        val iv = wrapped.copyOfRange(PUB_LEN, PUB_LEN + IV_LEN)
        val tag = wrapped.copyOfRange(PUB_LEN + IV_LEN, PUB_LEN + IV_LEN + TAG_LEN)
        val ct = wrapped.copyOfRange(PUB_LEN + IV_LEN + TAG_LEN, wrapped.size)

        val priv = X25519PrivateKeyParameters(recipientPrivateKey, 0)
        val recipientPubBytes = ByteArray(PUB_LEN).also { priv.generatePublicKey().encode(it, 0) }
        val shared = ByteArray(X25519Agreement.SECRET_SIZE)
        X25519Agreement().apply { init(priv) }
            .calculateAgreement(X25519PublicKeyParameters(ephPub, 0), shared, 0)
        val key = hkdfSha256(shared, salt = recipientPubBytes, info = INFO, length = 32)

        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(TAG_LEN * 8, iv))
        return cipher.doFinal(ct + tag)
    }

    private fun hkdfSha256(ikm: ByteArray, salt: ByteArray, info: ByteArray, length: Int): ByteArray {
        val out = ByteArray(length)
        HKDFBytesGenerator(SHA256Digest()).apply {
            init(HKDFParameters(ikm, salt, info))
            generateBytes(out, 0, length)
        }
        return out
    }
}
